File: BaseLib/src/main/java/com/hjc/baselib/utils/ClickUtils.java
package com.hjc.baselib.utils;

import java.util.HashMap;
import java.util.Map;

/**
 * @Author: HJC
 * @Date: 2019/1/7 11:34
 * @Description: 点击工具类
 */
public class ClickUtils {
    private static Map<String, Long> records = new HashMap<>();
    private static Map<String, Long> record2s = new HashMap<>();

    public static boolean isFastClick() {
        if (records.size() > 1000) {
            records.clear();
        }

        //本方法被调用的文件名和行号作为标记
        StackTraceElement ste = new Throwable().getStackTrace()[1];
        String key = ste.getFileName() + ste.getLineNumber();

        Long lastClickTime = records.get(key);
        long thisClickTime = System.currentTimeMillis();
        records.put(key, thisClickTime);
        if (lastClickTime == null) {
            lastClickTime = 0L;
        }
        long timeDuration = thisClickTime - lastClickTime;

        //如果两次点击间隔在500ms之内,则为快速点击
        return 0 < timeDuration && timeDuration < 500;
    }

    public static boolean isDoubleClick() {
        if (record2s.size() > 1000) {
            record2s.clear();
        }

        //本方法被调用的文件名和行号作为标记
        StackTraceElement ste = new Throwable().getStackTrace()[1];
        String key = ste.getFileName() + ste.getLineNumber();

        Long lastClickTime = record2s.get(key);
        long thisClickTime = System.currentTimeMillis();
        record2s.put(key, thisClickTime);
        if (lastClickTime == null) {
            lastClickTime = 0L;
        }
        long timeDuration = thisClickTime - lastClickTime;

        //如果两次点击间隔在2000ms之内,则为双击
        return 0 < timeDuration && timeDuration < 2000;
    }
}

File: BaseLib/src/main/java/com/hjc/baselib/utils/ClickUtils.kt
package com.hjc.baselib.utils

import java.util.*

/**
 * @Author: HJC
 * @Date: 2019/1/7 11:34
 * @Description: 点击工具类
 */
object ClickUtils {
    private val records: MutableMap<String, Long> = HashMap()
    private val record2s: MutableMap<String, Long> = HashMap()

    //本方法被调用的文件名和行号作为标记
    fun isFastClick(): Boolean {
        if (records.size > 1000) {
            records.clear()
        }

        //本方法被调用的文件名和行号作为标记
        val ste = Throwable().stackTrace[1]
        val key = ste.fileName + ste.lineNumber
        var lastClickTime = records[key]
        val thisClickTime = System.currentTimeMillis()
        records[key] = thisClickTime
        if (lastClickTime == null) {
            lastClickTime = 0L
        }

        val timeDuration = thisClickTime - lastClickTime
        //如果两次点击间隔在500ms之内,则为快速点击
        return timeDuration in 1..499
    }

    //本方法被调用的文件名和行号作为标记
    fun isDoubleClick(): Boolean {
        if (record2s.size > 1000) {
            record2s.clear()
        }
        //本方法被调用的文件名和行号作为标记
        val ste = Throwable().stackTrace[1]
        val key = ste.fileName + ste.lineNumber
        var lastClickTime = record2s[key]
        val thisClickTime = System.currentTimeMillis()
        record2s[key] = thisClickTime
        if (lastClickTime == null) {
            lastClickTime = 0L
        }
        val timeDuration = thisClickTime - lastClickTime
        //如果两次点击间隔在2000ms之内,则为双击
        return timeDuration in 1..1999
    }
}